service/matching/version_rule_helpers.go

```diff
@@ -191,49 +193,140 @@ func DeleteAssignmentRule(timestamp *hlc.Clock,
+// - The DAG of redirect rules must not contain a cycle
 func checkRedirectConditions(g *persistencepb.VersioningData, maxRRs int) error {
+	rules := g.GetRedirectRules()
+	if maxRRs > 0 && countActiveRR(rules) > maxRRs {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update would exceed number of redirect rules permitted in namespace dynamic config (%v/%v)", len(rules), maxRRs))
+	}
+	if isCyclic(rules) {
+		return serviceerror.NewFailedPrecondition("update would break acyclic requirement")
+	}
 	return nil
 }
 
 func countActiveRR(rules []*persistencepb.RedirectRule) int {
-	return 0
+	cnt := 0
+	for _, rule := range rules {
+		if rule.DeleteTimestamp == nil {
+			cnt++
+		}
+	}
+	return cnt
+}
+
+// isCyclic returns true if there is a cycle in the DAG of redirect rules.
+func isCyclic(rules []*persistencepb.RedirectRule) bool {
+	dag := makeEdgeMap(rules)
+	for node := range dag {
+		visited := make(map[string]bool)
+		inStack := make(map[string]bool)
+		if dfs(node, visited, inStack, dag) {
+			return true
+		}
+	}
+	return false
 }
 
-// trimRedirectRules attempts to trim the DAG of redirect rules. It returns the number of rules it was able to delete.
-func trimRedirectRules(rules []*persistencepb.RedirectRule) int {
-	return 0
+func makeEdgeMap(rules []*persistencepb.RedirectRule) map[string][]string {
+	ret := make(map[string][]string)
+	for _, rule := range rules {
+		src := rule.GetRule().GetSourceBuildId()
+		dst := rule.GetRule().GetTargetBuildId()
+		list, ok := ret[src]
+		if !ok {
+			list = make([]string, 0)
+		}
+		list = append(list, dst)
+		ret[src] = list
+	}
+	return ret
 }
 
-func isCyclic(rules []*persistencepb.RedirectRule) bool {
+func dfs(curr string, visited, inStack map[string]bool, nodes map[string][]string) bool {
+	if inStack[curr] {
+		return true
+	}
+	if visited[curr] {
+		return false
+	}
+	visited[curr] = true
+	inStack[curr] = true
+	for _, dst := range nodes[curr] {
+		if dfs(dst, visited, inStack, nodes) {
+			return true
+		}
+	}
+	inStack[curr] = false
 	return false
 }
 
 func InsertCompatibleRedirectRule(timestamp *hlc.Clock,
 	data *persistencepb.VersioningData,
 	req *workflowservice.UpdateWorkerVersioningRulesRequest_AddCompatibleBuildIdRedirectRule,
 	maxRRs int) (*persistencepb.VersioningData, error) {
-	return nil, nil
+	if data == nil {
+		data = &persistencepb.VersioningData{RedirectRules: make([]*persistencepb.RedirectRule, 0)}
+	} else {
+		data = common.CloneProto(data)
+	}
+	rule := req.GetRule()
+	src := rule.GetSourceBuildId()
+	rules := data.GetRedirectRules()
+	for _, r := range rules {
+		if r.GetDeleteTimestamp() != nil && r.GetRule().GetSourceBuildId() == src {
+			return nil, serviceerror.NewAlreadyExist(fmt.Sprintf(
+				"cannot insert: source %s already redirects to target %s",
+				src, r.GetRule().GetTargetBuildId(),
+			))
+		}
+	}
+	data.RedirectRules = slices.Insert(rules, 0, &persistencepb.RedirectRule{
```

Comment:
Need to reject if the src or target build ID is used in any versionset. Same for Replace. Same for assignment rules add/replace. same for Commit.

Reply:
ok. is the reason we don't describe that requirement in the API comments because eventually there should be no more version sets?